From: Mike Hommey <mh@glandium.org>
Date: Sun, 30 Aug 2015 18:35:43 +0900
Subject: Bug 1178266 - Link against libatomic when necessary

Some parts on various platforms needs the libatomic library with GCC5.
This patch implements a configure check that add the linker flag for
libatomic if needed automatically.

https://bugzilla.mozilla.org/show_bug.cgi?id=1178266
---
 mozilla/build/autoconf/toolchain.m4 | 26 ++++++++++++++++++++++++++
 mozilla/mfbt/moz.build              |  3 +++
 2 files changed, 29 insertions(+)

diff --git a/mozilla/build/autoconf/toolchain.m4 b/mozilla/build/autoconf/toolchain.m4
index 120a63c..90d8528 100644
--- a/mozilla/build/autoconf/toolchain.m4
+++ b/mozilla/build/autoconf/toolchain.m4
@@ -224,7 +224,33 @@ if test "$GNU_CXX"; then
     elif test "$ac_cv_cxx0x_headers_bug" = "yes"; then
         AC_MSG_ERROR([Your toolchain does not support C++0x/C++11 mode properly. Please upgrade your toolchain])
     fi
+
+    AC_CACHE_CHECK([whether 64-bits std::atomic requires -latomic],
+        ac_cv_needs_atomic,
+        AC_TRY_LINK(
+            [#include <cstdint>
+             #include <atomic>],
+            [ std::atomic<uint64_t> foo; foo = 1; ],
+            ac_cv_needs_atomic=no,
+            _SAVE_LIBS="$LIBS"
+            LIBS="$LIBS -latomic"
+            AC_TRY_LINK(
+                [#include <cstdint>
+                 #include <atomic>],
+                [ std::atomic<uint64_t> foo; foo = 1; ],
+                ac_cv_needs_atomic=yes,
+                ac_cv_needs_atomic="do not know; assuming no")
+            LIBS="$_SAVE_LIBS"
+        )
+    )
+    if test "$ac_cv_needs_atomic" = yes; then
+      MOZ_NEEDS_LIBATOMIC=1
+    else
+      MOZ_NEEDS_LIBATOMIC=
+    fi
+    AC_SUBST(MOZ_NEEDS_LIBATOMIC)
 fi
+
 if test -n "$CROSS_COMPILE"; then
     dnl When cross compile, we have no variable telling us what the host compiler is. Figure it out.
     cat > conftest.C <<EOF
diff --git a/mozilla/mfbt/moz.build b/mozilla/mfbt/moz.build
index 18ae476..3a91efe 100644
--- a/mozilla/mfbt/moz.build
+++ b/mozilla/mfbt/moz.build
@@ -122,3 +122,6 @@ if CONFIG['GNU_CXX']:
 if CONFIG['_MSC_VER']:
     # Error 4804 is "'>' : unsafe use of type 'bool' in operation"
     SOURCES['/mfbt/Compression.cpp'].flags += ['-wd4804']
+
+if CONFIG['MOZ_NEEDS_LIBATOMIC']:
+    OS_LIBS += ['atomic']
